demo_modules/slideshow/slideshow_server.js

```diff
@@ -24,68 +24,118 @@ limitations under the License.
-  };
+  }
 
   // MODULE DEFINTIONS
   class ImageServer {
     constructor(config) {
+      // The client loaded so far by the loading strategy.
+      this.content = [];
+
       // The load strategy for this run of the module.
       this.loadStrategy = parseServerLoadStrategy(config.load);
 
       // The display strategy for this run of the module.
-      this.displayStrategy = parseServerDisplayStrategy(config.display);
+      this.displayStrategy = parseServerDisplayStrategy(config.display, this);
+
+      // Caches used by the loading strategy when clipping images. The exact
+      // format of keys is determined by the loading strategy.
+      this.contentCache = new SizeLimitedCache(2**30);
     }
-    async willBeShownSoon() {
-      // Start the load strategy initing.
-      await Promise.all([
-        this.displayStrategy.init(),
-        this.loadStrategy.init().then(() => {
-          let fetchContent = (opt_paginationToken) => {
-            this.loadStrategy.loadMoreContent(opt_paginationToken).then((result) => {
-              this.displayStrategy.newContent(result.content);
-
-              if (result.hasMoreContent) {
-                fetchContent(result.paginationToken);
-              }
-            });
+    /**
+     * What to do when new content is downloaded.
+     */
+    handleNewContent(content) {
+      this.content.push(...random.shuffle(content));
+    }
+    /**
+     * Asks the load strategy to load some content and also starts a loop to
+     * load the rest of the content.
+     */
+    async startLoadingContent() {
+      await this.loadStrategy.init();
+      const firstResponse = await this.loadStrategy.loadMoreContent();
+      this.handleNewContent(firstResponse.content);
+      // Don't wait for any more content to download, but start downloading it.
+      this.loadRemainingContent(firstResponse.paginationToken);
+    }
+    /**
+     * Loads any remaining content starting with the specified paginationToken.
+     */
+    async loadRemainingContent(token) {
+      while(token) {
+        const response = await this.loadStrategy.loadMoreContent(token);
+        this.handleNewContent(response.content);
+        token = response.paginationToken;
+      }
+    }
+
+    async chooseContent(client) {
+      const ret = this.content.shift();
+      debug('Selected', ret, 'for client', client.offset);
+      this.content.push(ret);
+      // Ask the loading strategy to nab some metadata for this content item.
+      const metadata = await this.loadStrategy.metadataForContent(ret, this.contentCache);
+      if (metadata) {
+        debug(`Metadata: ${metadata.width} x ${metadata.height}`);
+        // If it chooses to response with some, we then ask the display
+        // strategy if it wants to clip this content.
+        const clippingRect = this.displayStrategy.clipRectForMetadata(metadata, client);
+        if (clippingRect) {
```

Comment:
This factoring feels like chooseContent is working too hard in the clipping case, but I don't have a better suggestion.

Reply:
Yeah.